compiler/rustc_middle/src/ty/relate.rs

```diff
@@ -273,7 +274,20 @@ impl<'tcx> Relate<'tcx> for ty::AliasTy<'tcx> {
         if a.def_id != b.def_id {
             Err(TypeError::ProjectionMismatched(expected_found(relation, a.def_id, b.def_id)))
         } else {
-            let args = relation.relate(a.args, b.args)?;
+            let args = match relation.tcx().def_kind(a.def_id) {
+                DefKind::OpaqueTy => relate_args_with_variances(
```

Comment:
shouldn't it be fine to use `variances_of` for all other aliases as well? at this point you could use `relation.relate_item_args` which avoids any lookup in `Equate` (and needs to call `variances_of` regardless if we're not in `Equate`)

Reply:
I don't think variances_of is impl'd for projections. I guess I could impl it, but that may have other perf implications.

Reply:
hmm, alternatively, instead of using the `DefKind`, have the `Ty::relate` provide the `AliasKind` to the substs relate.

do we need `ProjectionTy` to implement `Relate` itself, or would an inherent method `relate` which also takes the `AliasKind` be possible?

Reply:
> do we need ProjectionTy to implement Relate itself, or would an inherent method relate which also takes the AliasKind be possible?

This is pretty difficult, because in order to invoke a relation, you need something that implements `Self: ToTrace`. That requires `Self: Relate`. I guess I could poke a hole through with something like [`trace`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_infer/infer/at/struct.At.html#method.trace), but even then, it doesn't seem really ergonomic. Or I could separate out `ToTrace` from `Relate` or something...

This also makes this code significantly uglier:

https://github.com/rust-lang/rust/blob/56ada88e7e1f00e6c6e8a0957de295bc0e2d64f7/compiler/rustc_trait_selection/src/solve/alias_relate.rs#L171-L190

Since we can't just use the generic `eq` and `sub` methods on `EvalCtxt`.